Dockerfile

```diff
@@ -101,9 +106,13 @@ ENV JULIA_CPU_TARGET=x86_64;haswell;skylake;skylake-avx512;tigerlake
 # those to set up the ReefGuideAPI source code as a development package in the
 # shared @reefguide environment, pre-installing and precompiling dependencies.
 WORKDIR "${REEFGUIDE_SRC_DIR}"
-COPY ./Project.toml ./Project.toml
-COPY ./Manifest.toml ./Manifest.toml
-RUN julia --project=@reefguide -e 'using Pkg;  Pkg.instantiate(verbose=true)'
+COPY Project.toml Manifest.toml ./
+
+# Pre-download and cache MKL_jll dependency before general precompilation
+RUN julia -e 'using Pkg; Pkg.add("MKL_jll"); Pkg.instantiate(); using MKL_jll'
```

Comment:
Correct me if I'm wrong - `MKL_jll` was already a dependency listed in Project/Manifest, right? So the only condition under which this would be faster is if a) the user has already fully built the project b) they go to build again and c) the Project.toml and Manifest.toml has not changed. In that situation, wouldn't the whole pre-compilation already be skipped since the cache would apply to these three layers anyway 

```
COPY Project.toml Manifest.toml ./

# Pre-download and cache MKL_jll dependency before general precompilation
RUN julia -e 'using Pkg; Pkg.add("MKL_jll"); Pkg.instantiate(); using MKL_jll'

# Precompile Julia packages using BuildKit cache for better efficiency
RUN julia --project=@reefguide -e 'using Pkg; Pkg.instantiate(); Pkg.precompile()'
```

I'm not against pulling out specific 'base' dependencies that we think the image should always have, however, there are some risks

- what if the MKL_jll version changes - this is no longer tracked in the Project/Manifest, right? 
- why is this dependency special? Why not all the others? Is it just because it's slow?

If we do want to pull out a specific dependency, I think it makes more sense to do so before we copy the manifest files e.g. this order

```
RUN julia -e 'using Pkg; Pkg.add("MKL_jll"); Pkg.instantiate(); using MKL_jll'
COPY Project.toml Manifest.toml ./
RUN julia --project=@reefguide -e 'using Pkg; Pkg.instantiate(); Pkg.precompile()'
```

This way, you would never need to pre-comp this layer again even if manifest/project change. 

So in summary - I don't fully understand why this change actually reduces build times since a change in manifest/project will cause the whole thing to generally precompile again, and I have some concerns about adding hard coded dependencies which aren't version managed into the docker file.

Reply:
yes, I'm confused by this too. All dependencies should be defined by Project.toml

Reply:
@PeterBaker0 
**Correct me if I'm wrong - MKL_jll was already a dependency listed in Project/Manifest, right? So ..........**

MKL_jll is included through dependencies like LinearAlgebra in Project.toml, meaning it will be precompiled during Pkg.instantiate() and Pkg.precompile(). Docker caching will indeed optimise repeat builds when Project.toml and Manifest.toml remain unchanged by reusing layers and skipping full re-precompilation.

The explicit precompilation of MKL_jll primarily benefits initial build times and mitigates startup warnings. By precompiling MKL_jll separately, we handle its specific overhead upfront, reducing runtime costs and potential resource bottlenecks during the first build. However, in subsequent cached builds, this layer’s impact is minimized as Docker caching will efficiently skip redundant precompilations when dependencies haven’t changed.

**If we do want to pull out a specific dependency, I think it makes more sense to do so before we copy the manifest files**

We need MKL_jll to be installed with the exact version specified in Manifest.toml, so it’s important to copy both Project.toml and Manifest.toml files first. With this approach, even if Project.toml or Manifest.toml changes and triggers a full re-precompile, Docker will maintain a separate layer for MKL_jll. This separate layer can streamline subsequent builds, as MKL_jll will already be precompiled and cached independently of the other dependencies. This optimises for scenarios where MKL_jll remains unchanged while other dependencies are updated.


**what if the MKL_jll version changes - this is no longer tracked in the Project/Manifest, right?**
Again, I can see from the Manifest.toml file that LinearAlgebra is configured to use MKL. If this is not essential, we could switch to using OpenBLAS instead. By running this after copying the Manifest.toml file, we ensure that the version specified in Manifest.toml is used. I have removed Pkg.add to prevent it from pulling a different version, ensuring the version from Manifest.toml is used.

**why is this dependency special? Why not all the others? Is it just because it's slow?**

MKL_jll is singled out because it is a large and resource-intensive dependency with a known tendency to cause significant build-time delays and runtime warnings if not precompiled. By precompiling MKL_jll explicitly during the Docker build phase, the process prevents these warnings and reduces the load at runtime. Other dependencies, generally less demanding, don’t typically require this upfront handling, which is why they aren't singled out in the same way.

Reply:
> @PeterBaker0 **Correct me if I'm wrong - MKL_jll was already a dependency listed in Project/Manifest, right? So ..........**
> 
> MKL_jll is included through dependencies like LinearAlgebra in Project.toml, meaning it will be precompiled during Pkg.instantiate() and Pkg.precompile(). Docker caching will indeed optimise repeat builds when Project.toml and Manifest.toml remain unchanged by reusing layers and skipping full re-precompilation.
> 
> The explicit precompilation of MKL_jll primarily benefits initial build times and mitigates startup warnings. By precompiling MKL_jll separately, we handle its specific overhead upfront, reducing runtime costs and potential resource bottlenecks during the first build. However, in subsequent cached builds, this layer’s impact is minimized as Docker caching will efficiently skip redundant precompilations when dependencies haven’t changed.

What do you mean by "initial build times" - why would initial build times be reduced? You're not actually doing less in the build, right? 

"reducing runtime costs" - what runtime costs? Isn't this about decreasing build time? Are you saying the server starts faster?

"efficiently skip redundant precompilations when dependencies haven’t changed." I still don't understand why this is any different than how it was before? If the dependencies didn't change, then it would never have re-precompiled, right?



> 
> **If we do want to pull out a specific dependency, I think it makes more sense to do so before we copy the manifest files**
> 
> We need MKL_jll to be installed with the exact version specified in Manifest.toml, so it’s important to copy both Project.toml and Manifest.toml files first. With this approach, even if Project.toml or Manifest.toml changes and triggers a full re-precompile, Docker will maintain a separate layer for MKL_jll. This separate layer can streamline subsequent builds, as MKL_jll will already be precompiled and cached independently of the other dependencies. This optimises for scenarios where MKL_jll remains unchanged while other dependencies are updated.

I just don't think this is true. Docker cache layers will always be invalidated if the file hash of copied inputs changes. E.g. 

1) user adds package
2) manifest file etc change
3) this invalidates the copy layer and all layers below it 
4) MKL_jll layer needs to rebuild because step 3 invalidated this layer
5) other packages need to rebuild

> 
> **what if the MKL_jll version changes - this is no longer tracked in the Project/Manifest, right?** Again, I can see from the Manifest.toml file that LinearAlgebra is configured to use MKL. If this is not essential, we could switch to using OpenBLAS instead. By running this after copying the Manifest.toml file, we ensure that the version specified in Manifest.toml is used. I have removed Pkg.add to prevent it from pulling a different version, ensuring the version from Manifest.toml is used.

Sure - but then you will need to recompile MKL_jll anytime the Manifest or Project toml files change, which removes the benefit of pulling it out separately anyway?

> 
> **why is this dependency special? Why not all the others? Is it just because it's slow?**
> 
> MKL_jll is singled out because it is a large and resource-intensive dependency with a known tendency to cause significant build-time delays and runtime warnings if not precompiled. By precompiling MKL_jll explicitly during the Docker build phase, the process prevents these warnings and reduces the load at runtime. Other dependencies, generally less demanding, don’t typically require this upfront handling, which is why they aren't singled out in the same way.

Sure - but you would need to pin this version into the dockerfile itself and build it before you bring in project/manifest toml.

Reply:
My thoughts:

1. If MKL_jll does indeed contribute significantly to build times, then there would be benefit to precompiling it separately.

Unlike other approaches, say `conda` in Python, where a separate copy of the package is installed for each environment, Julia will share packages that are previously installed across environments _so long as the versions align_.

This would be beneficial in the case that the version installed by this line:

`RUN julia -e 'using Pkg; Pkg.add("MKL_jll"); Pkg.instantiate(); using MKL_jll'`

is indeed compatible with the version specified in the `Project.toml` and `Manifest.toml` files, as installation/compilation is effectively skipped given the previously installed package can be used instead.

But as soon as there is version misalignment (the line above initially installing an older version, and the ReefGuideAPI.jl package specifying a newer version), then all the benefits are lost.

So you need some way of running the line above again if/when the version of `MKL_jll` in the `Project.toml` file changes.

In the case the Project.toml file changes, but does not affect the `MKL_jll` dependency, then we would still get the improved build times.

Does this make sense?

Reply:
That's not how Docker cache works (from my understanding). 

Layer cache hash := hash(file system hash at start of layer, build command and files referenced in the build command)

If the project/manifest change at all, even just a new line or an extra whitespace, this hash is now different (since the build command/referenced files has changed). This will always invalidate all following layers. 

So afaik this "In the case the Project.toml file changes, but does not affect the MKL_jll dependency, then we would still get the improved build times." is not true. 

But I talked with @pascal082  and I am going to validate this is the case. 

I'm not unconvinced that the build times improved, just trying to understand why.

Reply:
Just confirming that the above explanation matches tested behaviour i.e. any change to dependencies (Manifest/Project files) results in the entire build occurring again including MKL_jll